scvi/external/sysvi/_module.py

```diff
@@ -0,0 +1,445 @@
+            x_m=generative_outputs["x_m"], x=x_true, x_v=generative_outputs["x_v"]
+        )
+        reconst_loss = reconst_loss_x
+
+        # Kl divergence on latent space
+        kl_divergence_z = self.prior.kl(
+            m_q=inference_outputs["z_m"],
+            v_q=inference_outputs["z_v"],
+            z=inference_outputs["z"],
+        )
+
+        def z_dist(z_x_m: torch.Tensor, z_y_m: torch.Tensor):
+            """MSE loss between standardised inputs with standardizer fitted on concatenation of both inputs"""
+            # Standardise data (jointly both z-s) before MSE calculation
+            z = torch.concat([z_x_m, z_y_m])
+            means = z.mean(dim=0, keepdim=True)
+            stds = z.std(dim=0, keepdim=True)
+
+            def standardize(x):
+                return (x - means) / stds
+
+            return torch.nn.MSELoss(reduction="none")(standardize(z_x_m), standardize(z_y_m)).sum(
+                dim=1
+            )
+
+        z_distance_cyc = z_dist(z_x_m=inference_outputs["z_m"], z_y_m=inference_outputs["z_cyc_m"])
+
+        loss = (
+            reconst_loss * reconstruction_weight
+            + kl_divergence_z * kl_weight
+            + z_distance_cyc * z_distance_cycle_weight
+        )
+
+        return LossOutput(
+            n_obs_minibatch=loss.shape[0],
+            loss=loss.mean(),
+            extra_metrics={
+                "reconstruction_loss": reconst_loss.mean(),
+                "kl_local": kl_divergence_z.mean(),
+                "z_distance_cycle": z_distance_cyc.mean(),
+            },
+        )
+
+    @staticmethod
+    def random_select_systems(system: torch.Tensor) -> torch.Tensor:
+        """For every cell randomly selects a new system that is different from the original system
+
+        Parameters
+        ----------
+        system
+            One hot encoded system information for each cell
+
+        Returns
+        -------
+        One hot encoding of newly selected system for each cell
+
+        """
+        # get available systems - those that are zero will become nonzero and vice versa
+        available_systems = 1 - system
+        # Get nonzero indices for each cell
+        row_indices, col_indices = torch.nonzero(available_systems, as_tuple=True)
+        col_pairs = col_indices.view(-1, system.shape[1] - 1)
+        # Select system for every cell from available systems
+        randomly_selected_indices = col_pairs.gather(
+            1,
+            torch.randint(
+                0,
+                system.shape[1] - 1,
+                size=(col_pairs.size(0), 1),
+                device=col_pairs.device,
+                dtype=col_pairs.dtype,
+            ),
+        )
+        new_tensor = torch.zeros_like(available_systems)
+        # generate system covariate tensor
+        new_tensor.scatter_(1, randomly_selected_indices, 1)
+
+        return new_tensor
+
+    def sample(self, *args, **kwargs):
+        raise NotImplementedError("")
```

Comment:
Why? It should be implemented here so more downstream functions work with sysVI.